### PR TITLE
sorter/leveldb(ticdc): separate write from table sorter

### PR DESCRIPTION
**Pull request by overvenus**
### What problem does this PR solve?

Issue Number: ref #4631

### What is changed and how it works?

It separate write from table sorter.

Note, for now it's dead code, will be enabled in later PRs.

### Check List 

Tests 

 - Unit test

### Release note 

```release-note
None
```

### CODE REVIEW

File: cdc/sorter/leveldb/message/task.go [hunk 1]
```diff
@@ -27,11 +28,20 @@ type Task struct {
 	UID     uint32
 	TableID uint64
 
+	// Input unsorted event for writers.
+	// Sorter.AddEntry -> writer.
+	InputEvent *model.PolymorphicEvent
+	// Latest resolved ts / commit ts for readers.
+	// writer -> reader
+	ReadTs ReadTs
```
Comment (sdojjy) [inline cdc/sorter/leveldb/message/task.go:36]: It makes the logic complex and unreadable to add so many fields in one struct, some fields are always nil in some cases.
And the `Task` is a part of `Message`,  The size will be larger and larger, maybe we should find a better way to extend the `Task` or `Message`.

File: cdc/sorter/leveldb/system/system.go [hunk 2]
```diff
@@ -68,10 +70,14 @@ func NewSystem(dir string, memPercentage float64, cfg *config.DBConfig) *System
 		WorkerNumber(cfg.Count).Build()
 	compactSystem, compactRouter := actor.NewSystemBuilder("sorter-compactor").
 		WorkerNumber(cfg.Count).Build()
+	writerSystem, writerRouter := actor.NewSystemBuilder("sorter-writer").
```
Comment (sdojjy) [inline cdc/sorter/leveldb/system/system.go:73]: we have  `sorter-db` and `sorter-writer`,   can you add some comments about `sorter-db`

File: cdc/sorter/leveldb/writer.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+		}
+	}
+
+	// Notify readers that there is something to read.
+	msg := actormsg.SorterMessage(message.Task{
+		UID:     w.uid,
+		TableID: w.tableID,
+		ReadTs: message.ReadTs{
+			MaxCommitTs:   maxCommitTs,
+			MaxResolvedTs: maxResolvedTs,
```
Comment (sdojjy) [inline cdc/sorter/leveldb/writer.go:97]: maybe the write task is not finished,  is it ok to send `maxResolvedTs` to the reader?

File: cdc/sorter/leveldb/writer.go [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+	"github.com/pingcap/tiflow/cdc/model"
+	"github.com/pingcap/tiflow/cdc/sorter/encoding"
+	"github.com/pingcap/tiflow/cdc/sorter/leveldb/message"
+	"github.com/pingcap/tiflow/pkg/actor"
+	actormsg "github.com/pingcap/tiflow/pkg/actor/message"
+	"github.com/prometheus/client_golang/prometheus"
+	"go.uber.org/zap"
+)
+
+// writer is a thin shim that batches, translates events into key vaule pairs
```
Comment (sdojjy) [inline cdc/sorter/leveldb/writer.go:29]:
```suggestion
// writer is a thin shim that batches, translates events into key-value pairs
```

File: cdc/sorter/leveldb/writer.go (hunk 3 above)
Comment (overvenus) [reply cdc/sorter/leveldb/writer.go:97]: Yes, as reader/writer send tasks to the same leveldb, so tasks are ordered.
ReadTs will trigger reader to take iterator from leveldb, it happens after writer sends writes to leveldb.
And before leveldb takes iterators, it flushed all pending writes, so iterators see all writes.

File: cdc/sorter/leveldb/writer.go [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+	// Notify readers that there is something to read.
+	msg := actormsg.SorterMessage(message.Task{
+		UID:     w.uid,
+		TableID: w.tableID,
+		ReadTs: message.ReadTs{
+			MaxCommitTs:   maxCommitTs,
+			MaxResolvedTs: maxResolvedTs,
+		},
+	})
+	// It's ok if send fails, as resolved ts events are received periodically.
```
Comment (zhaoxinyu) [inline cdc/sorter/leveldb/writer.go:100]: Is it possible that resolved ts events keep failing to be sent.

File: cdc/sorter/leveldb/message/task.go (hunk 1 above)
Comment (overvenus) [reply cdc/sorter/leveldb/message/task.go:36]: It's good point! Ideally tasks should be an enum type in rust, so tasks can be separated and more readable. Unfortunately, golang does not support it.

> And the Task is a part of Message, The size will be larger and larger, maybe we should find a better way to extend the Task or Message.

Maybe we should make Task be a pointer in Message? Though it makes memory allocation, not sure if it helps performance.

File: cdc/sorter/leveldb/writer.go (hunk 5 above)
Comment (overvenus) [reply cdc/sorter/leveldb/writer.go:100]: Only if reader is blocked for a long time. Eg. reading iterator becomes very slow. I think it rarely happens.

File: cdc/sorter/leveldb/message/task.go (hunk 1 above)
Comment (overvenus) [reply cdc/sorter/leveldb/message/task.go:36]: Also see, notes in #4631

File: cdc/sorter/leveldb/writer.go [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+		err := w.dbRouter.SendB(ctx, w.dbActorID, actormsg.SorterMessage(task))
+		if err != nil {
+			w.reportError("failed to send write request", err)
+			return false
+		}
+	}
+
+	// Notify reader that there is something to read.
+	//
+	// It's ok to noify reader immediately without waiting writes done,
```
Comment (zhaoxinyu) [inline cdc/sorter/leveldb/writer.go:93]:
```suggestion
	// It's ok to notify reader immediately without waiting writes done,
```

**Review by zhaoxinyu — APPROVED** [review]

**Review by sdojjy — APPROVED** [review]
lgtm

**Comment by zhaoxinyu** [comment]
TestMySQLSinkFlushResovledTs failed.

**Pull request merged**